### PR TITLE
heter for collective

### PR DESCRIPTION
**Pull request by kuizhiqing**
### PR types

New features

### PR changes

Others

### Describe

Heterogenous mix training represents the model training with heterogenous hardwares. Dygraph mode is only supported now. GPU/NPU/XPU are targeting devices for this prototype work.

The basic idea is very similar as the use of hierarchical communication topology. The low layer reduce the data within each node, while the upper layer reduce across all global nodes. 

image

### CODE REVIEW

File: paddle/fluid/imperative/gloo_context.cc [hunk 1]
```diff
@@ -176,6 +176,11 @@ void GLOOParallelContext::AllReduce(const framework::SelectedRows &src,
   }
 }
 
+void GLOOParallelContext::BroadCast(framework::Variable *src, int ring_id) {
```
Comment (sandyhouse) [inline paddle/fluid/imperative/gloo_context.cc]: Broadcast？Broadcast是一个单词。另外，这个接口没有实现，为什么还要添加这个接口呢？

File: paddle/fluid/imperative/gloo_context.h [hunk 2]
```diff
@@ -47,6 +47,8 @@ class GLOOParallelContext : public ParallelContext {
                          framework::Variable* dst, int ring_id,
                          bool use_calc_stream) override;
 
+  void BroadCast(framework::Variable* src, int ring_id) override;
```
Comment (sandyhouse) [inline paddle/fluid/imperative/gloo_context.h]: 1. 同上。
2. gloo接口为什么需要传入ring_id?

File: paddle/fluid/imperative/hccl_context.cc [hunk 3]
```diff
@@ -158,6 +158,29 @@ void HCCLParallelContext::AllReduceByStream(const framework::Variable &src,
   }
 }
 
+void HCCLParallelContext::BroadCast(framework::Variable *src, int ring_id) {
```
Comment (sandyhouse) [inline paddle/fluid/imperative/hccl_context.cc]: BroadCast -> Broadcast？

File: paddle/fluid/imperative/nccl_context.cc [hunk 4]
```diff
@@ -127,6 +135,20 @@ void NCCLParallelContext::AllReduceByStream(const framework::Variable &src,
   AllReduce(src, dst, strategy_, ring_id, use_calc_stream);
 }
 
+void NCCLParallelContext::BroadCast(framework::Variable *src, int ring_id) {
```
Comment (sandyhouse) [inline paddle/fluid/imperative/nccl_context.cc]: BroadCast -> Broadcast?

File: paddle/fluid/imperative/nccl_context.h [hunk 5]
```diff
@@ -60,6 +60,8 @@ class NCCLParallelContext : public ParallelContext {
                          framework::Variable* dst, int ring_id,
                          bool use_calc_stream) override;
 
+  void BroadCast(framework::Variable* src, int ring_id) override;
```
Comment (sandyhouse) [inline paddle/fluid/imperative/nccl_context.h]: 同上

File: paddle/fluid/imperative/parallel_context.h [hunk 6]
```diff
@@ -56,6 +56,8 @@ class ParallelContext {
                                  framework::Variable* dst, int ring_id,
                                  bool use_calc_stream) = 0;
 
+  virtual void BroadCast(framework::Variable* src, int ring_id) = 0;
```
Comment (sandyhouse) [inline paddle/fluid/imperative/parallel_context.h]: 同上

File: paddle/fluid/imperative/reducer.cc [hunk 7]
```diff
@@ -41,6 +42,9 @@ void Group::DivNRanks(const platform::DeviceContext &context, int64_t nranks) {
 #if defined(PADDLE_WITH_NCCL) || defined(PADDLE_WITH_RCCL)
     DivNRanks(tensor, nranks, context);
 #endif
+  } else if (platform::is_npu_place(tensor->place())) {
+    // TODO(kuizhiqing)
+    VLOG(4) << "divnrank for npu not support yet";
```
Comment (sandyhouse) [inline paddle/fluid/imperative/reducer.cc:47]: Abort?

**Review by sandyhouse — COMMENTED** [review]
看着都是动态图的，不能支持静态图吧？

**Review by sandyhouse — APPROVED** [review]
LGTM

**Review by XieYunshen — APPROVED** [review]

**Review by chenwhql — APPROVED** [review]

**Review by zhiqiu — APPROVED** [review]
LGTM for const_cast

**Review by fuyinno4 — APPROVED** [review]

**Review by raindrops2sea — APPROVED** [review]

**Pull request merged**